crates/ruff_linter/src/rules/ruff/rules/unnecessary_rounding.rs

```diff
@@ -0,0 +1,186 @@
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{AlwaysFixableViolation, Applicability, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{Arguments, Expr, ExprCall, ExprNumberLiteral, Number};
+use ruff_python_semantic::analyze::type_inference::{NumberLike, PythonType, ResolvedPythonType};
+use ruff_python_semantic::analyze::typing;
+use ruff_text_size::Ranged;
+
+/// ## What it does
+/// Checks for `round()` calls that have no effect on the input.
+///
+/// ## Why is this bad?
+/// The unnecessary call may make the code less readable.
+///
+/// ## Example
+///
+/// ```python
+/// a = round(1, 0)
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// a = 1
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct UnnecessaryRounding;
+
+impl AlwaysFixableViolation for UnnecessaryRounding {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Value being rounded is already an integer".to_string()
+    }
+
+    fn fix_title(&self) -> String {
+        "Remove unnecessary rounding".to_string()
+    }
+}
+
+/// RUF057
+pub(crate) fn unnecessary_rounding(checker: &mut Checker, call: &ExprCall) {
+    let arguments = &call.arguments;
+
+    if !checker.semantic().match_builtin_expr(&call.func, "round") {
+        return;
+    }
+
+    let Some((rounded, rounded_kind, ndigits_kind)) = rounded_and_ndigits(checker, arguments)
+    else {
+        return;
+    };
+
+    if matches!(ndigits_kind, NdigitsKind::Other) {
+        return;
+    }
+
+    let applicability = match (rounded_kind, ndigits_kind) {
+        (RoundedKind::LiteralInt | RoundedKind::ResolvedInt, _) => Applicability::Safe,
+
+        (
+            RoundedKind::LiteralFloat | RoundedKind::ResolvedFloat,
+            NdigitsKind::NotGiven | NdigitsKind::LiteralNone,
+        ) => Applicability::Safe,
+
+        (RoundedKind::InferredInt, _) => Applicability::Unsafe,
+
+        (RoundedKind::InferredFloat, NdigitsKind::NotGiven | NdigitsKind::LiteralNone) => {
+            Applicability::Unsafe
+        }
+
+        _ => return,
+    };
+
+    let edit = unwrap_round_call(checker, call, rounded);
+    let fix = Fix::applicable_edit(edit, applicability);
+
+    let diagnostic = Diagnostic::new(UnnecessaryRounding, call.range);
+
+    checker.diagnostics.push(diagnostic.with_fix(fix));
+}
+
+/// The type of the first argument to `round()`
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub(super) enum RoundedKind {
+    LiteralInt,
+    LiteralFloat,
+    ResolvedInt,
+    ResolvedFloat,
+    InferredInt,
+    InferredFloat,
```

Comment:
Can you tell me more about the motivation for distinguishing between `ResolvedInt` and `InferredInt`? I think this is also worth documenting because it's not obvious to readers why the distinction is necessary. 

It might be better to have more semantic names rather than the source  because both `typing::int` and `ResolvedPythonType` are doing type inference, which makes `InferredInt` rather confusing

Reply:
As a reader, I still wonder what the difference between `InferredInt` and `ResolvedInt` is. I think we should rename the variants or add documentation (the best option is probably to do both). 

Can you share an example where `InferredInt` returns `true` for something that shouldn't be an `int`? Is it because booleans are assignable to `int`: 

E.g. this is fine according to pyright

```python
def test(a: int) -> int:
    return a

test(True)  # Returns `True`
```

But `round(True)` evaluates to `1`

```
>>> round(True)
1
```

Are there more cases like this? If not, then how about renaming the `InferredInt` variant to `AnnotatedInt` or `Int` vs `LiteralInt` with a comment saying that one is equivalent to an `:int` annotation and the other `Literal[int]`?

Reply:
Thanks for adding some documentation. I did read it, but I'm not sure it's very helpful because it only documents what's already apparent from the code. What's unclear to a reader is why the distinction is important. The names are also somewhat confusing, as you mentioned in your last comment. For example, `LiteralInt` doesn't mean that it's a `Literal`; it could also be a complex expression. 

I still only have a limited understanding but my impression is that the variants for `RoundKind` are all about *how likely the expression is an `int`*. Or possibly whether the value is equivalent to `int` or only assignable to int. How about we introduce a new enum to `InferedType` with the variants `Equivalent` (it's exactly `int`), and `AssignableTo` (5 or `True`). `RoundKind` then has the variants `Int(InferredType)`, `Float(InferredType)`, and `Other` and `NdigitsKind` has `Absent`, `LiteralNone`, `Int(InferredType)`, and `Other`. 

We can then add documentation to where we construct the `InferredType` why e.g. `typing::is_int` is `InferredType::AssignableTo` and why `ResolvedPythonType` is `Equivalent`